File: python/insertionsort.py
# InsertionSort

# Similar to sorting playing cards
# It inserts a number into its correct position by shifting larger number to the right.


def insertionSort(list):
    N = len(list)

    for i in range(1, N):
        key = list[i]
        j = i - 1  # Compare with previous number
        while j >= 0 and key < list[j]:
            list[j + 1] = list[j]
            j -= 1  # Decrement j to compare with the element before
        list[j + 1] = key

        # We add 1 because if not it would be -1, or would be a position behind due to j-=1 in while loop.

    return list


listTest = [12, 11, 13, 5, 6]

print(insertionSort(listTest))

File: c/insertionsort.c
#include <stdio.h>

void insertionSort(int list[], int n);

int main(void)
{
    int listTest[7] = {64, 34, 25, 12, 22, 11, 90};
    int n = sizeof(listTest) / sizeof(listTest[0]);
    insertionSort(listTest, n);
}

void insertionSort(int list[], int n)
{
    for (int i = 1; i < n; i++)
    {
        int key = list[i];
        int j = i - 1;
        while ((j >= 0) && (key < list[j]))
        {
            list[j + 1] = list[j];
            j--;
        }
        list[j + 1] = key;
    }

    // Print Array
    for (int i = 0; i < n; i++)
    {
        printf("%i, ", list[i]);
    }
}
